### PR TITLE
Controlled shuffles check

### PR DESCRIPTION
**Pull request by yael1994**

### CODE REVIEW

**Comment by yael1994** [comment]
https://trello.com/c/J917NxDF/10-controlled-shuffles-pattern

File: hits_cpp/shufflesGenerator.hpp [hunk 1]
```diff
@@ -13,24 +13,36 @@ class MemeShuffler {
             maxShuffles = len;
         }
         this->_maxShuffles = maxShuffles;
+        this->_lenPatterns=len;
+        cout << "len pattern: " << this->_lenPatterns << endl;
+        this->_numPattern=-1;
```
Comment (shacharmo) [inline hits_cpp/shufflesGenerator.hpp]: Constant member should be initialized before constructor body.
see how `_source`, `_patterns` and `_currentShuffle` are initialized

File: hits_cpp/shufflesGenerator.hpp [hunk 2]
```diff
@@ -13,24 +13,36 @@ class MemeShuffler {
             maxShuffles = len;
         }
         this->_maxShuffles = maxShuffles;
+        this->_lenPatterns=len;
```
Comment (shacharmo) [inline hits_cpp/shufflesGenerator.hpp]: Following the coding style, adding spaces around `=`

File: hits_cpp/shufflesGenerator.hpp [hunk 3]
```diff
@@ -13,24 +13,36 @@ class MemeShuffler {
...
         while (iter != end) {
             meme.getRows().push_back(this->_source.getRows()[*iter]);
             iter++;
         }
+        //check that we don't change place of the same consensus:
+        for (int i = 0; i < this->_source.getRows().size(); i++){
+            auto max_source = max_element(this->_source.getRows()[i].begin(), this->_source.getRows()[i].end()) - this->_source.getRows()[i].begin();
+            auto max_shuffle = max_element(meme.getRows()[i].begin(), meme.getRows()[i].end()) - meme.getRows()[i].begin();
+            if (max_source == max_shuffle && (this->_lenPatterns - this->_numPattern) > (this->_maxShuffles - this->_currentShuffle)){
+                _numPattern++;
```
Comment (shacharmo) [inline hits_cpp/shufflesGenerator.hpp:39]: What happens if this is the last pattern?
What happens if no pattern matches, are we getting a random shuffles or the original PSSM?
If the original PSSM, what is the final rank of the PSSM 0 or 1.
We want the motif to be ranked as insignificant

**Review by shacharmo — CHANGES_REQUESTED** [review]

File: hits_cpp/shufflesGenerator.hpp (hunk 3 above)
Comment (yael1994) [reply hits_cpp/shufflesGenerator.hpp:39]: If all the shuffles do not do "full" shuffle so we take the last 5 shuffles of the patterns.

File: hits_cpp/shufflesGenerator.hpp (hunk 1 above)
Comment (shacharmo) [reply hits_cpp/shufflesGenerator.hpp]: Still not fixed, `_numPatterns` should also be initialized at start of construction, see `_currentShuffle`

File: hits_cpp/shufflesGenerator.hpp (hunk 3 above)
Comment (shacharmo) [reply hits_cpp/shufflesGenerator.hpp:39]: Tested: https://docs.google.com/document/d/1jR-LEQTyFpKO1Yf5QYNMSiIfRVy8qB8MC07U-zhkTuw
While rank is not always 0, the client authorized the solution

**Review by shacharmo — CHANGES_REQUESTED** [review]

**Review by shacharmo — APPROVED** [review]

**Pull request merged**